### ISSUE TITLE
Make travel time and edge weights independent

### ISSUE THREAD

**Issue by emiltin** [Project-OSRM/osrm-backend#77]
i think it will be imporant to be able to set 'attractiveness' on ways depending on certain tags.

this should be separate from the speed, since it's still important to have the correct speeds for each sections, so that a estimated travel time can be computed.

for example, we want to prioritize streets with bike tracks over streets without, even though the speed might be the same, or slightly lower. the same goes for green routes, which people often prefer over a big road, even though the speed might be similar or lower. 

one way it could be done simply is by hardcoding the tag combinations in osrm, but allowing the weight to be set in the speedprofile. something like:

   if(accesTag=='bicycle' &&  cycleway=='track')
        attractiveness = w.keyVals.Find("bike_track_attractiveness");    //value set in speedprofile, say 1.4
   else if(accesTag=='bicycle' &&  cycleway=='lane')
        attractiveness = w.keyVals.Find("bike_lane_attractiveness");    //value set in speedprofile, say 1.2
   else if(accesTag=='bicycle' &&  route != nil)
        attractiveness = w.keyVals.Find("bike_route_attractiveness");    //value set in speedprofile, say 1.5
  //other cases goes here....
  else
        attractiveness = 1.0;

**Comment by eleu** [comment]
I'd ask the same question as @emiltin about any updates on this one, agreeing that this would be a very useful feature.

**Comment by emiltin** [comment]
+1

**Comment by emiltin** [comment]
hi @TheMarex, what are the plans for separating travel time and impedance? it's quite akward having to always modify travel time to prioritized certain ways over other.

for example, we're implemeting green routes for bikes, which follows parks, lakes etc. currently the only way to do that is to modify the travel time, resulting in weird travel times being returned by OSRM. this means we  we need to fall back on hacks like ignoring what osrm provides and instead computing a generic travel time in the front-end as distance/average_bike_speed.

**Comment by TheMarex** [comment]
@emiltin This is certainly desirable, _but_ would mean a lot higher memory usage. An idea would be to hold auxiliary data for each edge out of main memory (for example street name, OSM id, height difference) that would not get used for routing but could be returned with the response to augment the route.

**Comment by emiltin** [comment]
yes it would require more memory, but only during preparing, right? also adding something like 8 bit for each edge might not be a huge increase compared to what's curently used?

not sure isee how storing auxilary data would solve the problem of  speed and impedance not being separated, but maybe i misunderstand you?

**Comment by emiltin** [comment]
you mean only keep one value for each edge (essentially the impedance) in memory and have anything else, like travel time, stored on disk, and fetch that after the route has been found? that  could work although i suppose the disk access would increase the reponse time significantly?

**Comment by emiltin** [comment]
any news on this?

**Comment by TheMarex** [comment]
This is a must-have feature for our bike routing. Just waiting for someone to free up to jump on this.

**Comment by emiltin** [comment]
ah great

**Comment by danpat** [comment]
In https://github.com/Project-OSRM/osrm-backend/issues/1844, @systemd found that capping line compression to values < 2^32 -1 (driven by the desire to use `short` instead of `int`) lead to only a surprisingly small increase in overall edge count after graph compression.

This hints that there are some bytes to be saved in the `EdgeBasedNode`, which we could re-purpose for time-independent weight, or perhaps just a scaling value (either a float, or a perhaps a https://en.wikipedia.org/wiki/Half-precision_floating-point_format).

It's possible that we could then implement this ticket with the only cost being a small number of additional edges.

**Comment by emiltin** [comment]
how many bits do you think would be available?

**Comment by MichalPP** [comment]
is it possible to store impedance on each edge and store speed where currently geometry is stored? 

the drawback would be that time returned by table request would not be time (but distance would be distance). 

the benefit would be possibility to store elevation profile, speed (or any extra info) separately, used only for post processing/display/text.

**Comment by TheMarex** [comment]
Congratulations @oxidase @jakepruitt and everyone that worked on #2399, we can now close this issue after 4 (in words FOUR) years of its creation.

**Comment by emiltin** [comment]
Hurray!

**Comment by jakepruitt** [comment]
Congrats everyone!

**Comment by danpat** [comment]
Anyone who has been watching this ticket - we've merge into master but feedback on this feature would be really helpful.  Anyone who has the time to test it out, please do and open tickets with any problems/confusion you encounter.

**Comment by nocallaghan** [comment]
@danpat I encountered a problem extracting a pbf file using weight_name set to distance. I've opened an issue: https://github.com/Project-OSRM/osrm-backend/issues/3638

**Comment by nocallaghan** [comment]
@danpat after #3640 was merged to master, this feature is working well using british-isles-latest.pbf. In all test cases so far, the shortest route was used while maintaining accurate travel times.
